src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,348 @@
+                (current_node_index, RestChild::None)
+            }
+        }
+    }
+
+    /// Attempt to move to the rest not visited child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the node [`Aabb`], `has_node` will become false.
+    fn move_rest(&mut self, rest_child: RestChild) {
+        match self.bvh.nodes[self.node_index] {
+            BvhNode::Node {
+                child_r_index,
+                child_l_index,
+                ..
+            } => match rest_child {
+                RestChild::Left => {
+                    self.node_index = child_l_index;
+                    self.has_node = true;
+                }
+                RestChild::Right => {
+                    self.node_index = child_r_index;
+                    self.has_node = true;
+                }
+                RestChild::None => {
+                    self.has_node = false;
+                }
+            },
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+            }
+        }
+    }
+}
+
+impl<'bvh, 'shape, T, const D: usize, Shape: Bounded<T, D>, const ASCENDING: bool> Iterator
+    for DistanceTraverseIterator<'bvh, 'shape, T, D, Shape, ASCENDING>
+where
+    T: Scalar + Copy + RealField,
+{
+    type Item = &'shape Shape;
+
+    fn next(&mut self) -> Option<&'shape Shape> {
+        loop {
+            if self.is_stack_empty() && !self.has_node {
+                // Completed traversal.
+                break;
+            }
+
+            if self.has_node {
+                // If we have any node, attempt to move to its nearest child.
+                let stack_info = if ASCENDING {
+                    self.move_nearest()
+                } else {
+                    self.move_furthest()
+                };
+                // Save curent node and farthest child
+                self.stack_push(stack_info)
+            } else {
+                // Go back up the stack and see if a node or leaf was pushed.
+                let (node_index, rest_child) = self.stack_pop();
+                self.node_index = node_index;
+                match self.bvh.nodes[self.node_index] {
+                    BvhNode::Node { .. } => {
+                        // If a node was pushed, now attempt to move to its rest farthest child.
+                        self.move_rest(rest_child);
+                    }
+                    BvhNode::Leaf { shape_index, .. } => {
+                        // We previously pushed a leaf node. This is the "visit" of the in-order traverse.
+                        // Next time we call `next()` we try to pop the stack again.
+                        return Some(&self.shapes[shape_index]);
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+// Copy of part of the BH testing in testbase.
+// TODO: Once iterators are part of the BoundingHierarchy trait we can move all this to testbase.
```

Comment:
Are you planning on implementing this TODO in a follow-up?

I'd ideally like to deduplicate this.